### PR TITLE
Add ability to retry functional tests which timeout

### PR DESCRIPTION
**Pull request by Naviabheeman**
This PR makes changes in the functional test framework to add the following features:

1. Retry tests which timeout
      After the initial run with all eligible tests the second run lists only the tests which fail on a timeout. The error code used to identify this has been changed and new status added to identify this.

2. Use a dynamic timeout value depending on the time taken to connect to a node
      When the test framework starts a node, it remembers the time taken to establish the RPC connection to it. Every RPC request to the node uses the time_to_connect*5 as the timeout. As this is very dynamic, according to the test environment we expect that the tests would not fail because of short timeout values.

### CODE REVIEW

File: test/functional/test_framework/key.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 import ctypes.util
 import hashlib
 
-ssl = ctypes.cdll.LoadLibrary(ctypes.util.find_library ('ssl') or 'libeay32')
+ssl = ctypes.cdll.LoadLibrary(ctypes.util.find_library ('ssl'))
```
Comment (azuchi) [inline test/functional/test_framework/key.py:16]: Why is this change necessary?

File: test/functional/test_framework/mininode.py [hunk 2]
```diff
@@ -236,6 +236,8 @@ def __init__(self):
         # The network services received from the peer
         self.nServices = 0
 
+        self.timeout = time_to_connect * 5
```
Comment (azuchi) [inline test/functional/test_framework/mininode.py:239]: Why are you multiplying by 5 here?

File: test/functional/test_framework/test_node.py [hunk 3]
```diff
@@ -231,7 +234,14 @@ def is_node_stopped(self):
...
     def wait_until_stopped(self, timeout=BITCOIND_PROC_WAIT_TIMEOUT):
-        wait_until(self.is_node_stopped, timeout=timeout)
+        i = 0
+        while i < 2:
+            try:
+                i = i + 1
+                wait_until(self.is_node_stopped, timeout=timeout)
+            except TimeoutError as e:
+                pass
+
```
Comment (azuchi) [inline test/functional/test_framework/test_node.py:244]: What is the intention behind this addition of logic?

File: test/functional/test_framework/test_node.py [hunk 4]
```diff
@@ -245,7 +255,7 @@ def assert_start_raises_init_error(self, extra_args=None, expected_msg=None, mat
              tempfile.NamedTemporaryFile(dir=self.stdout_dir, delete=False) as log_stdout:
             try:
                 self.start(extra_args, stdout=log_stdout, stderr=log_stderr, *args, **kwargs)
-                self.wait_for_rpc_connection()
+                timeout = self.wait_for_rpc_connection()
```
Comment (azuchi) [inline test/functional/test_framework/test_node.py:258]: This `timeout` variable used? If not, `wait_for_rpc_connection`  does not need to return this.

File: test/functional/p2p_unrequested_blocks.py [hunk 5]
```diff
@@ -284,11 +284,11 @@ def run_test(self):
             # Only wait a short while so the test doesn't take forever if we do get
             # disconnected
             test_node.sync_with_ping(timeout=1)
-        except AssertionError:
+        except TimeoutError:
```
Comment (azuchi) [inline test/functional/p2p_unrequested_blocks.py:287]: If it is an AssertionError, does it matter if it is no longer processed?

File: test/functional/test_framework/mininode.py (hunk 2 above)
Comment (Naviabheeman) [reply test/functional/test_framework/mininode.py:239]: 5 is just a random number, there is no reason behind it. I want to make the timeout long enough to wait for network and wallet to sync

File: test/functional/test_framework/test_node.py (hunk 3 above)
Comment (Naviabheeman) [reply test/functional/test_framework/test_node.py:244]: this was to double the timeout period for stopping a node. Some tests were failing because the node did not stop within the given time.

File: test/functional/test_framework/test_node.py (hunk 4 above)
Comment (Naviabheeman) [reply test/functional/test_framework/test_node.py:258]: right! its not used. will fix it.

File: test/functional/test_framework/key.py (hunk 1 above)
Comment (Naviabheeman) [reply test/functional/test_framework/key.py:16]: this is leftover from my test environment. it usually complaints about library libeay32 not found when I run functional tests. I remove it as a shortcut and continue texting. it shouldn't be in the repo. will fix it.

File: test/functional/test_framework/mininode.py (hunk 2 above)
Comment (azuchi) [reply test/functional/test_framework/mininode.py:239]: can you  add this intention as a comment?

File: test/functional/p2p_unrequested_blocks.py (hunk 5 above)
Comment (Naviabheeman) [reply test/functional/p2p_unrequested_blocks.py:287]: When the test is run with `AssertionError` the test exit code seems to be different and the status is
```
 2023-01-20T11:36:31.303000Z TestFramework (WARNING): Timeout. None
2023-01-20T11:36:31.360000Z TestFramework (INFO): Stopping nodes
2023-01-20T11:36:32.191000Z TestFramework (INFO): Cleaning up /var/folders/0y/rhp0p3vx2rj70m8zxhgpw1c80000gn/T/test3ss97r3u on exit
2023-01-20T11:36:32.192000Z TestFramework (INFO): Test timeout
``` 

with `TimeoutError` it is different:
```
2023-01-20T11:38:29.427000Z TestFramework (INFO): Successfully synced nodes 1 and 0
2023-01-20T11:38:29.483000Z TestFramework (INFO): Stopping nodes
2023-01-20T11:38:30.307000Z TestFramework (INFO): Cleaning up /var/folders/0y/rhp0p3vx2rj70m8zxhgpw1c80000gn/T/testvaocqx1o on exit
2023-01-20T11:38:30.307000Z TestFramework (INFO): Tests successful
```

**Review by azuchi — APPROVED** [review]
